File: pipeline/seo.py
# pipeline/seo.py
"""Generate SEO-optimized metadata for the YouTube video.
Creates a click-bait title and a description with tags using OpenAI.
"""
import json
from openai import OpenAI
from pathlib import Path
from .config import CHATGPT_API_KEY

client = OpenAI(api_key=CHATGPT_API_KEY)

def generate_metadata(script_path: Path = Path("output/script.txt"), output_path: Path = Path("output/metadata.json")):
    """Read the script and generate title/description."""
    if not script_path.exists():
        print(f"Script not found at {script_path}, skipping metadata generation.")
        return

    script_excerpt = script_path.read_text(encoding="utf-8")[:1000]  # First 1000 chars for context
    
    prompt = f"""
    Based on the following motivational script excerpt, generate:
    1. A high-CTR, clickbait YouTube title (max 60 chars) specifically targeted at a US audience (American English idioms, emotionally resonant).
    2. An SEO-optimized video description (max 200 words) for US viewers, including keywords like #motivation, #discipline, #success.
    
    Script Excerpt:
    "{script_excerpt}..."
    
    Return the result as valid JSON with keys: "title", "description".
    """

    response = client.chat.completions.create(
        model="gpt-4o-mini",
        messages=[{"role": "system", "content": prompt}],
        temperature=0.7,
    )
    
    content = response.choices[0].message.content.strip()
    try:
        # cleanup markdown code blocks if present
        if content.startswith("```json"):
            content = content.strip("```json").strip("```")
        elif content.startswith("```"):
            content = content.strip("```")
            
        data = json.loads(content)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        with open(output_path, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2)
        print(f"Metadata saved to {output_path}")
    except json.JSONDecodeError:
        print("Failed to parse SEO metadata JSON. Raw output:")
        print(content)

if __name__ == "__main__":
    generate_metadata()

File: pipeline_js/seo.js
// pipeline_js/seo.js
const fs = require('fs');
const path = require('path');
const { OpenAI } = require('openai');
const config = require('./config');

const client = new OpenAI({
    apiKey: config.CHATGPT_API_KEY,
});

async function generateSeo(
    scriptPath = path.join(config.OUTPUT_DIR, "script.txt"),
    outputPath = path.join(config.OUTPUT_DIR, "metadata.txt")
) {
    console.log("🏷️ Generating SEO metadata (JS)...");

    if (!fs.existsSync(scriptPath)) {
        console.error("❌ Script file not found!");
        return null;
    }

    const script = fs.readFileSync(scriptPath, 'utf8');

    const prompt = `Generate highly optimized SEO metadata for a motivational YouTube video based on this script.

Script Summary:
${script.substring(0, 2000)}...

Provide:
1. Five high-CTR TITLES (max 70 chars)
2. One deep, emotional DESCRIPTION (include keywords, timestamps placeholder, and call-to-actions)
3. 20-30 TAGS categorized by impact

Format your response as valid JSON:
{
  "titles": ["", ""],
  "description": "",
  "tags": ""
}
`;

    try {
        const response = await client.chat.completions.create({
            model: "gpt-4o-mini",
            messages: [{ role: "user", content: prompt }],
            temperature: 0.7,
            max_tokens: 1500,
            response_format: { type: "json_object" }
        });

        const metadata = response.choices[0].message.content.trim();

        const dir = path.dirname(outputPath);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
        }

        fs.writeFileSync(outputPath, metadata, 'utf8');
        console.log(`✓ SEO metadata saved to ${outputPath}`);
        return outputPath;
    } catch (error) {
        console.error("❌ Error generating SEO:", error.message);
        throw error;
    }
}

if (require.main === module) {
    generateSeo().catch(err => process.exit(1));
}

module.exports = { generateSeo };
